odo/backends/url.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import print_function, division, absolute_import
+
+import os
+import uuid
+import zlib
+import re
+
+from contextlib import contextmanager
+
+import requests
+from toolz import memoize
+
+from .. import (discover, CSV, resource, append, convert, drop, Temp, JSON,
+                JSONLines, chunks)
+
+from multipledispatch import MDNotImplementedError
+
+from .text import TextFile
+
+from ..compatibility import urlparse
+from ..utils import tmpfile, ext, sample, filter_kwargs
+
+
+class _URL(object):
+    """ Parent class for data accessed through ``URLs``
+
+    Parameters
+    ----------
+
+    url : str
+        full url to file
+    chunk_size : int (default 1024)
+        Size of chunks streamed into memory
+    decode_unicode : bool (default False)
+        If True, content will be decoded using the best available encoding based on the response.
+
+     Examples
+    --------
+
+    >>> from odo import URL, CSV
+    >>> u = URL(CSV)('http://foo.com/myfile.csv')
+
+    Normally create through resource uris
+
+    >>> data = resource('http://foo.com/myfile.csv')
+    >>> data.url
+    'http://foo.com/myfile.csv'
+    >>> data.filename
+    'myfile.csv'
+    """
+    def __init__(self, url, chunk_size=1024, decode_unicode=False, *args,
+                 **kwargs):
+
+        self.subtype.__init__(self, url, *args, **kwargs)
+
+        self.url = url
+        self.chunk_size = chunk_size
+        self.decode_unicode = decode_unicode
+        self.filename = os.path.basename(urlparse(url).path)
+
+
+def URL(cls):
+    return type('URL(%s)' % cls.__name__, (_URL, cls), {'subtype': cls})
+
+URL.__doc__ = _URL.__doc__
+URL = memoize(URL) # what's happening here
+
+@sample.register((URL(CSV), URL(JSONLines)))
+@contextmanager
+def sample_url_line_delimited(data, lines=5):
+    """Get a size `length` sample from an URL CSV or URL line-delimited JSON.
+
+    Parameters
+    ----------
+    data : URL(CSV)
+        A hosted CSV
+    lines : int, optional, default ``5``
+        Number of lines to read into memory
+    """
+
+    r = requests.get(data.url, stream=True)
+
+    raw = []
+    if r.status_code == 200:
```

Comment:
Do you think we need a specialized exception or is `requests.raise_for_status()` ok.

```
In [24]: url = "http://foo.com/myfile.csv"

In [25]: r = requests.get(url, stream=True)

In [26]: r.raise_for_status()
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
<ipython-input-26-79bf9267f7bd> in <module>()
----> 1 r.raise_for_status()

/Users/quasiben/anaconda/envs/blz_dev/lib/python2.7/site-packages/requests/models.pyc in raise_for_status(self)
    829
    830         if http_error_msg:
--> 831             raise HTTPError(http_error_msg, response=self)
    832
    833     def close(self):

HTTPError: 404 Client Error: Not Found